File: day3/day3.py
file = open('input.txt', 'r')
#file = open('example.txt', 'r')
Lines = file.readlines()
total_sum = 0

signList = []
dotList = []
digits = []
numbers = set()
y=0

def checkNeighborDigits(sign, dotList):
    neighborDigitList = []
    if (sign[0]-1,sign[1]-1) not in dotList:
        neighborDigitList.append((sign[0]-1,sign[1]-1))
    if (sign[0] , sign[1] - 1) not in dotList:
        neighborDigitList.append((sign[0], sign[1] - 1))
    if (sign[0] + 1, sign[1] - 1) not in dotList:
        neighborDigitList.append((sign[0] + 1, sign[1] - 1))
    if (sign[0] - 1, sign[1]) not in dotList:
        neighborDigitList.append((sign[0] - 1, sign[1]))
    if (sign[0] + 1, sign[1]) not in dotList:
        neighborDigitList.append((sign[0] + 1, sign[1]))
    if (sign[0] - 1, sign[1] + 1) not in dotList:
        neighborDigitList.append((sign[0] - 1, sign[1] + 1))
    if (sign[0] , sign[1] + 1) not in dotList:
        neighborDigitList.append((sign[0], sign[1] + 1))
    if (sign[0] + 1, sign[1] + 1) not in dotList:
        neighborDigitList.append((sign[0] + 1, sign[1] + 1))
    return neighborDigitList

'''
467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..
'''
def parseNumbers(neihborDigitList):
    numbers = set()
    y=0
    for line in Lines:
        x = 0
        for indice in range(len(line)-1):
            if (x,y) in neighborDigitList:
                number = 0
                digits = 0
                if line[x-2].isdigit():
                    number += int(line[x-2]) * 100
                    digits += 1
                if line[x-1].isdigit():
                    number += int(line[x-1]) * 10
                    digits += 1
                else:
                    number = 0
                    digits = 0
                if line[x].isdigit():
                    number += int(line[x])
                    digits+=1
                if line[x+1].isdigit() and digits < 3:
                    number *= 10
                    number += int(line[x+1])
                    digits += 1
                if line[x+2].isdigit() and digits < 3:
                    number *= 10
                    number += int(line[x+2])
                numbers.add(number)
            x += 1
        y += 1
    return numbers


for line in Lines:
    x=0
    for char in line:
        if not(char.isdigit()) and char != "." and x != len(line)-1:
            signList.append((x,y))
        if char == ".":
            dotList.append((x,y))
        x += 1
    y += 1

for sign in signList:
    neighborDigitList = (checkNeighborDigits(sign, dotList))
    numbers = numbers.union(parseNumbers(neighborDigitList))
print(sum(numbers))

File: day3/day3.java
package day3;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class day3 {
    static Logger logger = Logger.getLogger(day3.class.getName());
    static final String EXAMPLE_FILE_PATH = "day3/example.txt";
    static final String ACTUAL_FILE_PATH = "day3/input.txt";
    static int sum = 0;

    public static List<Integer> getNeighborNumbers(MapObject sign, Map map){
        Integer x = sign.getLocationStartIndexX();
        Integer y = sign.getLocationStartIndexY();
        return map.getMapObjects().stream()
                .filter(a -> a.getObjType().equals(MapObject.ObjTypeEnum.NUMBER))
                .filter(number -> {
                    boolean yRange = Math.max(y - 1, number.getLocationStartIndexY()) == Math.min(number.getLocationStartIndexY(), y + 1);
                    boolean xRange = Math.max(x - number.getLength(), number.getLocationStartIndexX()) == Math.min(number.getLocationStartIndexX(), x + 1);
                    return yRange && xRange;
                }).mapToInt(MapObject::getValue).boxed().collect(Collectors.toList());

    }

    public static Integer getGearRatios(MapObject sign, Map map){
        Integer x = sign.getLocationStartIndexX();
        Integer y = sign.getLocationStartIndexY();
        List <Integer> neighbors = map.getMapObjects().stream()
                .filter(a -> a.getObjType().equals(MapObject.ObjTypeEnum.NUMBER))
                .filter(number -> {
                    boolean yRange = Math.max(y - 1, number.getLocationStartIndexY()) == Math.min(number.getLocationStartIndexY(), y + 1);
                    boolean xRange = Math.max(x - number.getLength(), number.getLocationStartIndexX()) == Math.min(number.getLocationStartIndexX(), x + 1);
                    return yRange && xRange;
                }).mapToInt(MapObject::getValue).boxed().collect(Collectors.toList());

        if(neighbors.size()>1){
            return neighbors.stream().reduce(1,(a,b) -> a*b);
        }
        return 0;

    }

    public static void main(String[] args) {
        List<String> lines = utils.FileParseUtil.readLinesFromFile(ACTUAL_FILE_PATH, logger);
        Map map = new Map(lines.getFirst().length(),lines.size(),new ArrayList<>());
        List<Integer> numbersToSum = new ArrayList<>();

        for (int i = 0; i < lines.size(); i++) {
            Pattern patternNumber = Pattern.compile("\\d+");

            Matcher matcherNumber = patternNumber.matcher(lines.get(i));

            Pattern patternSign = Pattern.compile("[^\\d.]");

            Matcher matcherSign = patternSign.matcher(lines.get(i));

            // Loop through all matches
            while (matcherSign.find()) {
                int index = matcherSign.start();
                map.getMapObjects().add(new MapObject(MapObject.ObjTypeEnum.SIGN,
                        index,i,1,-1));
            }

            // Loop through all matches
            while (matcherNumber.find()) {
                String number = matcherNumber.group();
                int index = matcherNumber.start();
                map.getMapObjects().add(new MapObject(MapObject.ObjTypeEnum.NUMBER,
                        index,i,number.length(),Integer.parseInt(number)));
            }
        }
        for (MapObject mapObject : map.getMapObjects()){
            if (mapObject.getObjType().equals(MapObject.ObjTypeEnum.SIGN)){
                numbersToSum.add(getGearRatios(mapObject,map));
            }
        }
        System.out.println(numbersToSum.stream().reduce(0, Integer::sum));

    }
}
